cirq-core/cirq/qis/clifford_tableau.py

```diff
@@ -176,6 +179,67 @@ def _str_full_(self) -> str:
 
         return string
 
+    def then(self, second: 'CliffordTableau') -> 'CliffordTableau':
+        """Returns a composed CliffordTableau of this tableau and the second tableau.
+
+        Then composed tableau is equal to (up to global phase) the composed
+        unitary operation of the two tableaux, i.e. equivalent to applying the unitary
+        operation of this CliffordTableau then applying the second one.
+
+        Args:
+            second: The second CliffordTableau to compose with.
+
+        Returns:
+            The composed CliffordTableau.
+
+        Raises:
+            TypeError: If the type of second is not CliffordTableau.
+            ValueError: If the number of qubits in the second tableau mismatch with
+                this tableau.
+        """
+        if not isinstance(second, CliffordTableau):
+            raise TypeError("The type for second tableau must be the CliffordTableau type")
+        if self.n != second.n:
+            raise ValueError(
+                f"Mismatched number of qubits of two tableaux: {self.n} vs {second.n}."
+            )
+
+        # Convert the underlying data type from bool to int for easier numerical computation.
+        m1 = self.matrix().astype(int)
+        m2 = second.matrix().astype(int)
+
+        p1 = self.rs.astype(int)
+        p2 = second.rs.astype(int)
+
+        merged_m = np.mod(m1.dot(m2), 2)
+        phase = np.mod(p1 + m1.dot(p2), 2)
+
+        # we need more phase correction for expanding Y to XZ and swapping Z_iX_i order.
+        for k in range(2 * self.n):
+            swap_phase = 0  # value betwen 0 and 3 representing [1, i, -1, -i] respectively.
+            prev_row_sum = np.zeros([2 * self.n])
+            swap_phase += np.sum(m1[k, : self.n] * m1[k, self.n :])  # Y gate => iXZ
+            for i, v in enumerate(m1[k]):
+                if v == 0:
+                    continue
+                swap_phase += np.sum(m2[i, : self.n] * m2[i, self.n :])  # Y gate => iXZ
```

Comment:
Well, it cannot. Note `if v==0: continue`, not every row will count.

Reply:
Sorry, I meant that we could do this before the outer loop:

```
        num_ys1 = [np.sum(m1[k, : self.n] * m1[k, self.n :]) for k in range(2 * self.n)]
        num_ys2 = [np.sum(m2[k, : self.n] * m2[k, self.n:]) for k in range(2 * self.n)]
```

And then here we could refer to that "cache": 

```
      swap_phase += num_ys2[i]  # Y gate => iXZ
```

Reply:
Not sure how much it saves - it depends on the number of columns that have 1 in the tableau...and for those who have all 0 columns (would that be a valid symplectic matrix? e.g. [[+IX +IX], [+IZ +IZ]]?), it would be more expensive...

Reply:
I see your point now. It can be further vectorized into:
```
num_ys1 = np.sum(m1[:, : self.n] * m1[:, self.n :], axis=1)
num_ys2 = np.sum(m2[:, : self.n] * m2[:, self.n :], axis=1)
```
I guess it worths doing it.